File: typescript/src/generator.ts
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * TypeScript rewrite derived from NPGenerator V2.0.2.
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba / Programmer: Masaya Kiwada
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import {
  BlockConstraint,
  KindOfAnswer,
  SolverMethod,
  Status,
  sqrt,
} from "./core.js";
import { JavaRandom } from "./random.js";
import { addNumber, answer } from "./solver.js";

export class Generator {
  readonly seedMethod = new SolverMethod();
  method = new SolverMethod();
  forbidden = -1;

  private readonly hiddenList: number[];
  private readonly hintList: number[];
  private readonly groups: number[][];
  private readonly seedStatus: Status;
  private readonly status: Status;
  private readonly groupStatus: Status;
  private readonly original: Status;
  private readonly trial: Status;
  private readonly candidates: number[] = [];
  private seed: number[];

  constructor(
    private readonly numSize: number,
    private readonly hint: number[],
    private readonly hidden: number[],
    private readonly block: BlockConstraint,
    private readonly random: JavaRandom,
    initialSeed?: number[],
  ) {
    this.seed = initialSeed === undefined ? [] : [...initialSeed];
    this.hiddenList = hidden.flatMap((value, index) => value !== 0 ? [index] : []);
    this.hintList = hint.flatMap((value, index) => value !== 0 ? [index] : []);
    this.seedStatus = new Status(numSize, block);
    this.status = new Status(numSize, block);
    this.groupStatus = new Status(numSize, block);
    this.original = new Status(numSize, block);
    this.trial = new Status(numSize, block);
    this.seedMethod.localization = true;
    this.seedMethod.nakedPair = true;
    this.seedMethod.hiddenPair = true;

    let groupCount = sqrt(this.hintList.length);
    if (groupCount === 0) groupCount = 1;
    const remainder = this.hintList.length % groupCount;
    const unit = Math.trunc(this.hintList.length / groupCount);
    let cursor = 0;
    this.groups = Array.from({ length: groupCount }, (_, index) => {
      const count = unit + (index < remainder ? 1 : 0);
      const group = this.hintList.slice(cursor, cursor + count);
      cursor += count;
      return group;
    });
  }

  setMethod(method: SolverMethod): void {
    this.method = method;
  }

  setForbidden(forbidden: number): void {
    this.forbidden = forbidden;
  }

  private generateSeedSub(): number[] | undefined {
    const state = this.seedStatus;
    state.clear();
    for (let cell = 0; cell < this.hidden.length; cell++) {
      const value = this.hidden[cell]!;
      if (value > 0) addNumber(state, cell, value);
    }
    answer(state, this.seedMethod);
    for (let cell = 0; cell < this.hidden.length; cell++) {
      if (!state.isEmptyCell(cell)) continue;
      const candidateCount = state.getCandCountOfCell(cell);
      if (candidateCount === 0 || state.isNoAnswer()) return undefined;
      const randomIndex = this.random.nextInt(candidateCount);
      let value = state.getNthCandOfCell(cell, randomIndex);
      if (this.hint[cell] !== 0 && value === this.forbidden) {
        if (candidateCount === 1) return undefined;
        value = state.getNthCandOfCell(cell, (randomIndex + 1) % candidateCount);
      }
      addNumber(state, cell, value);
      answer(state, this.seedMethod);
    }
    if (this.forbidden > 0) {
      for (const cell of this.hintList) {
        if (state.getCell(cell) === this.forbidden) return undefined;
      }
    }
    return state.getSpaceCount() === 0 ? [...state.cells()] : undefined;
  }

  private generateSeed(): number[] | undefined {
    for (let failed = 0; failed <= 100; failed++) {
      const value = this.generateSeedSub();
      if (value !== undefined) return value;
    }
    return undefined;
  }

  private fitsHidden(cells: ArrayLike<number>): boolean {
    return this.hiddenList.every((cell) => cells[cell] === this.hidden[cell]);
  }

  generate(): number[] | undefined {
    const generatedSeed = this.generateSeed();
    if (generatedSeed === undefined) return undefined;
    this.seed = generatedSeed;

    const problem = Array(this.hidden.length).fill(0) as number[];
    for (const cell of this.hintList) problem[cell] = this.seed[cell]!;

    const state = this.status;
    state.clear();
    state.setUniqueMethod(this.method.unique);
    for (let cell = 0; cell < problem.length; cell++) {
      if (problem[cell]! > 0) addNumber(state, cell, problem[cell]!);
    }
    answer(state, this.method);
    if (state.getKindOfAnswer() === KindOfAnswer.UNIQUE_ANSWER) return problem;

    this.random.shuffle(this.hintList);
    let zero = state.getSpaceCount();
    if (zero === 0) return problem;
    let yet = true;
    while (yet) {
      yet = false;
      for (let groupIndex = 0; groupIndex < this.groups.length; groupIndex++) {
        const group = this.groups[groupIndex]!;
        const groupStatus = this.groupStatus;
        groupStatus.clear();
        groupStatus.setUniqueMethod(this.method.unique);
        for (let otherIndex = 0; otherIndex < this.groups.length; otherIndex++) {
          if (groupIndex === otherIndex) continue;
          for (const cell of this.groups[otherIndex]!) {
            addNumber(groupStatus, cell, problem[cell]!);
          }
        }
        for (const cell of group) {
          if (this.hidden[cell] !== 0) continue;
          const previous = problem[cell]!;
          problem[cell] = 0;
          const original = this.original;
          original.copyStatusToThis(groupStatus);
          for (const other of group) {
            if (other !== cell) addNumber(original, other, problem[other]!);
          }
          if (original.getCandCountOfCell(cell) <= 1) {
            problem[cell] = previous;
            continue;
          }
          const candidates = this.candidates;
          candidates.length = 0;
          const candidateMask = original.getCandidateMask(cell);
          for (let number = 1; number <= this.numSize; number++) {
            if ((candidateMask & (1 << (number - 1))) !== 0) candidates.push(number);
          }
          if (!candidates.includes(previous)) candidates.push(previous);
          problem[cell] = previous;
          const previousIndex = candidates.indexOf(previous);
          for (let offset = 1; offset < candidates.length; offset++) {
            const next = candidates[(previousIndex + offset) % candidates.length]!;
            if (next === previous || next === this.forbidden) continue;
            const trial = this.trial;
            trial.copyStatusToThis(original);
            addNumber(trial, cell, next);
            answer(trial, this.method);
            const space = trial.getSpaceCount();
            if (!trial.isNoAnswer() && this.fitsHidden(trial.cells()) && zero > space) {
              zero = space;
              problem[cell] = next;
              if (zero === 0) return problem;
              yet = true;
              break;
            }
          }
          if (yet) break;
        }
      }
    }
    return zero === 0 ? problem : undefined;
  }
}

File: rust/src/generator.rs
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * Rust rewrite derived from NPGenerator V2.0.2.
 *
 * Number Place Generator Version 2.0
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba
 * Programmer: Masaya Kiwada
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use std::sync::Arc;

use crate::model::{AnswerKind, BlockConstraint, SolverMethod, Status};
use crate::random::JavaRandom;
use crate::solver::{add_number, answer_in_place};

pub struct Generator<'a> {
    size: usize,
    hint: Vec<i32>,
    hidden: Vec<i32>,
    seed: Vec<i32>,
    hidden_list: Vec<usize>,
    hint_list: Vec<usize>,
    groups: Vec<Vec<usize>>,
    random: &'a mut JavaRandom,
    seed_method: SolverMethod,
    method: SolverMethod,
    forbidden: i32,
    seed_status: Status,
    solve_status: Status,
    group_status: Status,
    original: Status,
    trial: Status,
}

impl<'a> Generator<'a> {
    pub fn new(
        size: usize,
        hint: Vec<i32>,
        hidden: Vec<i32>,
        block: Arc<BlockConstraint>,
        random: &'a mut JavaRandom,
    ) -> Self {
        Self::new_with_seed(size, hint, hidden, block, random, None)
    }

    pub fn new_with_seed(
        size: usize,
        hint: Vec<i32>,
        hidden: Vec<i32>,
        block: Arc<BlockConstraint>,
        random: &'a mut JavaRandom,
        seed: Option<Vec<i32>>,
    ) -> Self {
        let hidden_list = hidden
            .iter()
            .enumerate()
            .filter_map(|(index, &value)| (value != 0).then_some(index))
            .collect();
        let hint_list: Vec<_> = hint
            .iter()
            .enumerate()
            .filter_map(|(index, &value)| (value != 0).then_some(index))
            .collect();
        let mut group_count = (hint_list.len() as f64).sqrt() as usize;
        if group_count == 0 {
            group_count = 1;
        }
        let remainder = hint_list.len() % group_count;
        let unit = hint_list.len() / group_count;
        let mut groups = Vec::new();
        let mut cursor = 0;
        for i in 0..group_count {
            let count = unit + usize::from(i < remainder);
            groups.push(hint_list[cursor..cursor + count].to_vec());
            cursor += count;
        }
        let mut seed_method = SolverMethod::none();
        seed_method.localization = true;
        seed_method.naked_pair = true;
        seed_method.hidden_pair = true;
        let seed_status = Status::new(size, Arc::clone(&block));
        let solve_status = Status::new(size, Arc::clone(&block));
        let group_status = Status::new(size, Arc::clone(&block));
        let original = Status::new(size, Arc::clone(&block));
        let trial = Status::new(size, Arc::clone(&block));
        Self {
            size,
            hint,
            hidden,
            seed: seed.unwrap_or_default(),
            hidden_list,
            hint_list,
            groups,
            random,
            seed_method,
            method: SolverMethod::none(),
            forbidden: -1,
            seed_status,
            solve_status,
            group_status,
            original,
            trial,
        }
    }

    pub fn set_method(&mut self, method: SolverMethod) {
        self.method = method;
    }

    pub fn set_forbidden(&mut self, forbidden: i32) {
        self.forbidden = forbidden;
    }

    fn generate_seed_sub(&mut self) -> Option<Vec<i32>> {
        let status = &mut self.seed_status;
        status.clear();
        for (cell, &n) in self.hidden.iter().enumerate() {
            if n > 0 {
                add_number(status, cell, n);
            }
        }
        answer_in_place(status, self.seed_method);
        for cell in 0..self.hidden.len() {
            if status.cell[cell] != 0 {
                continue;
            }
            let count = status.cand_count_cell(cell) as i32;
            if count == 0 || status.is_no_answer() {
                return None;
            }
            let random_index = self.random.next_int(count);
            let mut n = status.nth_candidate(cell, random_index);
            if self.hint[cell] != 0 && n == self.forbidden {
                if count == 1 {
                    return None;
                }
                n = status.nth_candidate(cell, (random_index + 1) % count);
            }
            add_number(status, cell, n);
            answer_in_place(status, self.seed_method);
        }
        if self.forbidden > 0
            && self
                .hint_list
                .iter()
                .any(|&cell| status.cell[cell] == self.forbidden)
        {
            return None;
        }
        (status.space_count == 0).then(|| status.cell.clone())
    }

    fn generate_seed(&mut self) -> Option<Vec<i32>> {
        for _ in 0..=100 {
            if let Some(seed) = self.generate_seed_sub() {
                return Some(seed);
            }
        }
        None
    }

    fn fits_hidden(&self, cell: &[i32]) -> bool {
        self.hidden_list
            .iter()
            .all(|&index| cell[index] == self.hidden[index])
    }

    pub fn generate(&mut self) -> Option<Vec<i32>> {
        self.seed = self.generate_seed()?;
        let mut problem = vec![0; self.hidden.len()];
        for &cell in &self.hint_list {
            problem[cell] = self.seed[cell];
        }
        let status = &mut self.solve_status;
        status.clear();
        status.unique = self.method.unique;
        for (cell, &n) in problem.iter().enumerate() {
            if n > 0 {
                add_number(status, cell, n);
            }
        }
        answer_in_place(status, self.method);
        if status.kind == AnswerKind::Unique {
            return Some(problem);
        }

        self.random.shuffle(&mut self.hint_list);
        let mut zero = status.space_count;
        if zero == 0 {
            return Some(problem);
        }
        let mut candidates = Vec::with_capacity(self.size);
        let mut yet = true;
        while yet {
            yet = false;
            for group_index in 0..self.groups.len() {
                self.group_status.clear();
                self.group_status.unique = self.method.unique;
                for (other_index, other_group) in self.groups.iter().enumerate() {
                    if group_index != other_index {
                        for &cell in other_group {
                            add_number(&mut self.group_status, cell, problem[cell]);
                        }
                    }
                }
                let group_len = self.groups[group_index].len();
                for cell_index in 0..group_len {
                    let cell = self.groups[group_index][cell_index];
                    if self.hidden[cell] != 0 {
                        continue;
                    }
                    let previous = problem[cell];
                    problem[cell] = 0;
                    self.original.copy_from_status(&self.group_status);
                    for other_index in 0..group_len {
                        let other = self.groups[group_index][other_index];
                        if other != cell {
                            add_number(&mut self.original, other, problem[other]);
                        }
                    }
                    if self.original.cand_count_cell(cell) <= 1 {
                        problem[cell] = previous;
                        continue;
                    }
                    candidates.clear();
                    for n in 1..=self.size as i32 {
                        if self.original.is_cand(cell, n) {
                            candidates.push(n);
                        }
                    }
                    if !candidates.contains(&previous) {
                        candidates.push(previous);
                    }
                    let previous_index = candidates
                        .iter()
                        .position(|&n| n == previous)
                        .expect("previous candidate must be present");
                    candidates.rotate_left(previous_index);
                    candidates.remove(0);
                    problem[cell] = previous;
                    for candidate_index in 0..candidates.len() {
                        let next = candidates[candidate_index];
                        if next == previous || next == self.forbidden {
                            continue;
                        }
                        self.trial.copy_from_status(&self.original);
                        add_number(&mut self.trial, cell, next);
                        answer_in_place(&mut self.trial, self.method);
                        let space = self.trial.space_count;
                        if !self.trial.is_no_answer()
                            && self.fits_hidden(&self.trial.cell)
                            && zero > space
                        {
                            zero = space;
                            problem[cell] = next;
                            if zero == 0 {
                                return Some(problem);
                            }
                            yet = true;
                            break;
                        }
                    }
                    if yet {
                        break;
                    }
                }
            }
        }
        (zero == 0).then_some(problem)
    }
}
